autoblocks/_impl/global_state.py

```diff
@@ -25,10 +29,19 @@ def init() -> None:
         daemon=True,
     )
     background_thread.start()
-
+    signal.signal(signal.SIGTERM, _on_unexpected_exit)
+    signal.signal(signal.SIGINT, _on_unexpected_exit)
     _started = True
 
 
+def _on_unexpected_exit(signum, frame):
+    if _loop and _loop.is_running():
+        log.info(f"Attempting to gracefully stop event loop. Received signal {signum}")
+        _loop.stop()
+        _loop.close()
+    exit(0)
+
+
```

Comment:
@nicolewhite my initial stab at attempting a graceful exit  - Struggling to come up with something to test this

Reply:
Where are we waiting for the futures to complete?

Reply:
Maybe something like this can be added

```python
def _on_exit(signum, frame):
    if _loop and _loop.is_running():
        log.info(f"Attempting to gracefully stop event loop. Received signal {signum}")
        _loop.stop()
        pending = asyncio.all_tasks()
        _loop.run_until_complete(asyncio.gather(*pending))
    exit(0)
```

Bit stuck on this part

Reply:
as for testing, you could try putting together a script that has evaluators that just sleep for like 5 seconds and run that script with `subprocess`, kill the process, then check the  events exist by fetching them via the public api. i'd add it to tests/e2e

```
import subprocess

proc = subprocess.Popen(...)
proc.send_signal(signal.SIGTERM)

# wait for events to exist in public api
```

or something, haven't tried it

Reply:
Added a test 👍